pkg/apis/batch/types.go

```diff
@@ -238,6 +238,20 @@ type CronJobList struct {
 	Items []CronJob
 }
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CronJobManualInstantiation represents a manual triggering of a CronJob.
+type CronJobManualInstantiation struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// CreatedJob is the Job created in the instantiation request.
+	CreatedJob Job
```

Comment:
This is not needed, I'd expect the `/instantiate` endpoint to return you that `Job` instance.

Reply:
I did it this way because I could not figure out how to circumvent the API machinery in such a way to allow me to return a different type of object than the one specified in the subresource's resource.

Generally speaking, I could not find any other examples of doing what I did, which is to have a subresource return a completely different type of object from its resource. I mostly referenced `pod/logs` when writing this code, which does not return a `Pod` instance.

Reply:
See OpenShift's [build/instantiate](https://github.com/openshift/origin/blob/9f48098688b100e4d5f7b889818eab826eedb774/pkg/build/registry/buildconfiginstantiate/rest.go#L60-L75). The api machinery is not restrictive in this way. You should not have any problems with returning a different object.

Reply:
Here's the problem I'm having with accommodating this request. It wasn't a problem with the type of the object, but rather the version.

I rewrote the genclient line as:
```
// +genclient:method=Instantiate,verb=create,subresource=instantiate,input=k8s.io/api/batch/v1beta1.CronJobManualRequest,result=k8s.io/api/batch/v1.Job
```
...which generated:
```
// Instantiate takes the representation of a cronJobManualRequest and creates it.  Returns the server's representation of the job, and an error, if there is any.
func (c *cronJobs) Instantiate(cronJobName string, cronJobManualRequest *v1beta1.CronJobManualRequest) (result *batch_v1.Job, err error) {
    result = &batch_v1.Job{}
    err = c.client.Post().
        Namespace(c.ns).
        Resource("cronjobs").
        Name(cronJobName).
        SubResource("instantiate").
        Body(cronJobManualRequest).
        Do().
        Into(result)
    return
}
```
...and that seems all fine.

However, hitting the API server seems to fail at the `Into(result)` point with:
```
{
    "kind": "Status",
    "apiVersion": "v1",
    "metadata": {},
    "status": "Failure",
    "message": "no kind \"Job\" is registered for version \"batch/v1beta1\"",
    "code": 500
}
```

It doesn't seem to be able to mix API versions. The request works up to this point; the job is created as expected, but `Into()` seems to be trying to materialize a Job object of kind `batch/v1beta1`.